File: baml_language/sdks/nodejs/bridge_nodejs/typescript_src/typemap.ts
// typemap.ts — runtime BamlTypeMap, the Node analog of
// sdks/python/src/baml_core/typemap.py.
//
// Codegen emits `_typemap.ts` with `BamlTypeMap.fromLazyEntries({ classes,
// enums, typeAliases })` where each entry is `[modulePath, attrName]`. The
// root `index.ts` calls `setTypeMap(_TYPE_MAP)` at import time. Resolution is
// lazy: `getClass(fqn)` does `require(modulePath)[attrName]` on first lookup
// and memoizes. (The decode-side walk that consumes this lands in Phase 5.)

import { BamlError } from './errors';

export type LazyEntry = [string, string]; // [modulePath, attrName]

// Hard-coded stdlib reverse-overrides. Mirrors _STDLIB_REVERSE_OVERRIDES in
// baml_core/typemap.py. Keys are `${modulePath}::${exportName}` of the native
// class identities the codegen-emitted re-exports point at. Phase 4/5 seed the
// real native identities (BamlImage/…/BamlStream); Phase 2 leaves it empty and
// documents the intent.
const _STDLIB_REVERSE_OVERRIDES: Map<string, string> = new Map([
    // ["@boundaryml/baml-core::BamlImage", "baml.media.Image"], — wired in Phase 4/5
]);

export class BamlTypeMap {
    private classLazy = new Map<string, LazyEntry>();
    private enumLazy = new Map<string, LazyEntry>();
    private aliasLazy = new Map<string, LazyEntry>();
    private classCache = new Map<string, unknown>();
    private enumCache = new Map<string, unknown>();
    private aliasCache = new Map<string, unknown>();
    private reverse: Map<string, string> = new Map(_STDLIB_REVERSE_OVERRIDES);

    static fromLazyEntries(args: {
        classes: Record<string, LazyEntry>;
        enums: Record<string, LazyEntry>;
        typeAliases: Record<string, LazyEntry>;
    }): BamlTypeMap {
        const m = new BamlTypeMap();
        for (const [fqn, le] of Object.entries(args.classes)) m.classLazy.set(fqn, le);
        for (const [fqn, le] of Object.entries(args.enums)) m.enumLazy.set(fqn, le);
        for (const [fqn, le] of Object.entries(args.typeAliases)) m.aliasLazy.set(fqn, le);
        for (const [fqn, [mp, attr]] of Object.entries(args.classes)) {
            const k = `${mp}::${attr}`;
            if (!m.reverse.has(k)) m.reverse.set(k, fqn);
        }
        for (const [fqn, [mp, attr]] of Object.entries(args.enums)) {
            const k = `${mp}::${attr}`;
            if (!m.reverse.has(k)) m.reverse.set(k, fqn);
        }
        return m;
    }

    private _resolve(
        fqn: string,
        lazy: Map<string, LazyEntry>,
        cache: Map<string, unknown>,
        kind: string,
    ): unknown {
        if (cache.has(fqn)) return cache.get(fqn);
        const entry = lazy.get(fqn);
        if (entry === undefined) throw new BamlError(`Unknown ${kind} FQN ${fqn}`);
        const [modulePath, attr] = entry;
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const mod = require(modulePath);
        const resolved = mod[attr];
        if (resolved === undefined) {
            throw new BamlError(`Could not resolve ${fqn} → ${modulePath}.${attr}`);
        }
        cache.set(fqn, resolved);
        return resolved;
    }

    getClass(fqn: string): unknown {
        return this._resolve(fqn, this.classLazy, this.classCache, 'class');
    }

    getEnum(fqn: string): unknown {
        return this._resolve(fqn, this.enumLazy, this.enumCache, 'enum');
    }

    getTypeAlias(fqn: string): unknown {
        return this._resolve(fqn, this.aliasLazy, this.aliasCache, 'type alias');
    }

    // Walk the prototype chain. Returns "" if no match. Phase 5 refines this
    // for class identity. (Provisional — see 10a-todo-items.md Phase D.)
    jsTypeToBamlType(cls: unknown): string {
        let cur: unknown = cls;
        while (cur != null) {
            const name = (cur as { name?: string }).name;
            const mod = (cur as { __bamlModulePath?: string }).__bamlModulePath;
            if (mod && name) {
                const fqn = this.reverse.get(`${mod}::${name}`);
                if (fqn !== undefined) return fqn;
            }
            cur = Object.getPrototypeOf(cur);
        }
        return '';
    }

    warm(): void {
        for (const k of this.classLazy.keys()) this.getClass(k);
        for (const k of this.enumLazy.keys()) this.getEnum(k);
        for (const k of this.aliasLazy.keys()) this.getTypeAlias(k);
    }
}

let _TYPE_MAP = new BamlTypeMap();
export function setTypeMap(m: BamlTypeMap): void {
    _TYPE_MAP = m;
}
export function getTypeMap(): BamlTypeMap {
    return _TYPE_MAP;
}

File: baml_language/sdks/python/src/baml_bridge/typemap.py
"""Process-global FQN → Python class registry (25a2 §4.1).

The typemap is codegen-emitted: each SDK ships a `baml_sdk/_typemap.py`
that constructs a `BamlTypeMap` from three literal dicts of
`FQN → (module_path, attr_name)` lazy entries, and the SDK's root
`__init__.py` installs it via `set_type_map(_TYPE_MAP)`. Resolution
happens on first `get_class(fqn)` call via `importlib.import_module +
getattr`, then memoizes.
"""

from __future__ import annotations
import importlib
from typing import Dict, Tuple, Type

from .errors import BamlError

_LazyEntry = Tuple[str, str]  # (module_path, attr_name)

# Hardcoded reverse-map seeds for the five stdlib re-exports whose
# Python class identity sits at baml_bridge.baml_py.BamlImage etc.
# instead of at baml_sdk.baml.media.Image. The forward emit path
# (sdkgen_python_pydantic2's `media_reexport_rust_name`) keeps its own
# hardcoded match arms; the duplication is small enough that a
# shared source of truth isn't worth a cross-crate constant.
_STDLIB_REVERSE_OVERRIDES: Dict[Tuple[str, str], str] = {
    ("baml_bridge.baml_py", "BamlImage"): "baml.media.Image",
    ("baml_bridge.baml_py", "BamlAudio"): "baml.media.Audio",
    ("baml_bridge.baml_py", "BamlVideo"): "baml.media.Video",
    ("baml_bridge.baml_py", "BamlPdf"): "baml.media.Pdf",
    # `BamlStream` is re-exported from `baml_bridge` but defined in
    # `baml_bridge._stream`; `__module__` reflects the defining module.
    ("baml_bridge._stream", "BamlStream"): "ai.stream.Stream",
    ("baml_bridge._function_spec", "BamlFunctionSpec"): "ai.FunctionSpec",
}


class BamlTypeMap:
    __slots__ = (
        # Lazy entries — codegen-emitted, resolved on first lookup.
        "_class_lazy",
        "_enum_lazy",
        "_alias_lazy",
        # Resolved cache — populated by first successful lazy resolution.
        "_class_cache",
        "_enum_cache",
        "_alias_cache",
        # Reverse map: (module, qualname) → engine FQN. Populated from
        # forward entries in `from_lazy_entries`, seeded with stdlib
        # PyO3-identity overrides. `py_type_to_baml_type` walks
        # `cls.__mro__` against this dict.
        "_reverse",
    )

    def __init__(self) -> None:
        self._class_lazy: Dict[str, _LazyEntry] = {}
        self._enum_lazy: Dict[str, _LazyEntry] = {}
        self._alias_lazy: Dict[str, _LazyEntry] = {}
        self._class_cache: Dict[str, Type] = {}
        self._enum_cache: Dict[str, Type] = {}
        self._alias_cache: Dict[str, object] = {}
        # Seed with the stdlib identity overrides every typemap needs
        # (BamlImage at baml_bridge.baml_py → "baml.media.Image", etc.).
        # Forward entries added by `from_lazy_entries` populate more
        # keys on top.
        self._reverse: Dict[Tuple[str, str], str] = dict(_STDLIB_REVERSE_OVERRIDES)

    @classmethod
    def from_lazy_entries(
        cls,
        classes: Dict[str, _LazyEntry],
        enums: Dict[str, _LazyEntry],
        type_aliases: Dict[str, _LazyEntry],
    ) -> "BamlTypeMap":
        m = cls()
        m._class_lazy = dict(classes)
        m._enum_lazy = dict(enums)
        m._alias_lazy = dict(type_aliases)
        # Derive (module, attr) → FQN from forward entries.
        # `setdefault` lets stdlib seeds (populated in __init__) win
        # on collision — for stdlib classes both the user-facing
        # re-export key AND the PyO3 identity key end up in the
        # reverse map; lookups on either resolve to the same FQN.
        for fqn, (mp, attr) in classes.items():
            m._reverse.setdefault((mp, attr), fqn)
        for fqn, (mp, attr) in enums.items():
            m._reverse.setdefault((mp, attr), fqn)
        # Type aliases generally don't appear as `type(value)`; skip.
        return m

    # — lookup (lazy fallback) —

    def get_class(self, fqn: str) -> Type:
        cached = self._class_cache.get(fqn)
        if cached is not None:
            return cached
        entry = self._class_lazy.get(fqn)
        if entry is None:
            raise BamlError(
                f"Unknown class FQN {fqn!r}; codegen did not emit a "
                "typemap entry (or codegen drift left it stale)"
            )
        module_path, attr = entry
        try:
            module = importlib.import_module(module_path)
            cls = getattr(module, attr)
        except (ImportError, AttributeError) as exc:
            raise BamlError(
                f"Could not resolve {fqn!r} → {module_path}.{attr}: {exc}"
            ) from exc
        self._class_cache[fqn] = cls
        return cls

    def get_enum(self, fqn: str) -> Type:
        cached = self._enum_cache.get(fqn)
        if cached is not None:
            return cached
        entry = self._enum_lazy.get(fqn)
        if entry is None:
            raise BamlError(f"Unknown enum FQN {fqn!r}")
        module_path, attr = entry
        try:
            module = importlib.import_module(module_path)
            cls = getattr(module, attr)
        except (ImportError, AttributeError) as exc:
            raise BamlError(
                f"Could not resolve enum {fqn!r} → {module_path}.{attr}: {exc}"
            ) from exc
        self._enum_cache[fqn] = cls
        return cls

    def get_type_alias(self, fqn: str) -> object:
        cached = self._alias_cache.get(fqn)
        if cached is not None:
            return cached
        entry = self._alias_lazy.get(fqn)
        if entry is None:
            raise BamlError(f"Unknown type alias FQN {fqn!r}")
        module_path, attr = entry
        try:
            module = importlib.import_module(module_path)
            alias = getattr(module, attr)
        except (ImportError, AttributeError) as exc:
            raise BamlError(
                f"Could not resolve alias {fqn!r} → {module_path}.{attr}: {exc}"
            ) from exc
        self._alias_cache[fqn] = alias
        return alias

    # — reverse lookup (replaces _baml_type_name ClassVar pathway) —

    def py_type_to_baml_type(self, cls: type) -> str:
        """Reverse lookup: Python class → engine FQN. Walks the MRO so
        user subclasses of generated classes resolve to the parent's
        FQN (matching today's ClassVar inheritance). Returns `""` for
        any class not in the typemap — informational-only field on the
        wire, same as 25b's `_derive_baml_fqn` fallback."""
        for c in cls.__mro__:
            fqn = self._reverse.get((c.__module__, c.__qualname__))
            if fqn is not None:
                return fqn
        return ""

    def warm(self) -> None:
        for fqn in list(self._class_lazy):
            self.get_class(fqn)
        for fqn in list(self._enum_lazy):
            self.get_enum(fqn)
        for fqn in list(self._alias_lazy):
            self.get_type_alias(fqn)


_TYPE_MAP = BamlTypeMap()


def set_type_map(m: BamlTypeMap) -> None:
    global _TYPE_MAP
    _TYPE_MAP = m


def get_type_map() -> BamlTypeMap:
    return _TYPE_MAP
